### ISSUE TITLE
C# style analyzer: invoke statics from the type on which they're declared

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#59094]
**Brief description:**

Consider:
```C#
Derived.M();

class Base
{
    public static void M() { }
}

class Derived : Base { }
```
This compiles fine, with no warnings, no analyzers firing, even though we're invoking `M` from the derived type while it was defined on the base.

This capability stemming back to C# 1.0 causes issues for API design.  Consider wanting to add a method like `ArgumentException.ThrowIfNullOrEmpty`.  It looks well and good, until you realize that all of these then also compile:
```C#
ArgumentNullException.ThrowIfNullOrEmpty(arg);
ArgumentOutOfRangeException.ThrowIfNullOrEmpty(arg);
RegexParserException.ThrowIfNullOrEmpty(arg);
CultureNotFoundException.ThrowIfNullOrEmpty(arg);
EncoderFallbackException.ThrowIfNullOrEmpty(arg);
```
Etc.

It'd be great if there were an analyzer / style rule that urged developers to use the type on which statics are declared when invoking those static methods.

**Languages applicable:**

C#

**Code example that the analyzer should report:**

See description above.

**Additional information:**

n/a

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at [Documentation for IDE CodeStyle analyzers](https://github.com/dotnet/roslyn/blob/main/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md).

**Comment by datvm** [comment]
Sorry if I am mistaken but I thought this is already exist at least back in the .NET FW 3.5? I remember receiving warning/suggestion that I should use `WebRequest.CreateHttp` instead of `HttpWebRequest.CreateHttp` (bad example now because this one is deprecated and I am not aware of any similar API).

![image](https://user-images.githubusercontent.com/6388546/151279638-df244397-0562-4424-8512-aa9ddcb32552.png)

 The code rule is [IDE0002](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0002)

**Comment by stephentoub** [comment]
Thanks. And interesting. That is not firing for me, even after I enable everything. Maybe it's an issue with the preview build in using? 

But more importantly, how is this "simplification"? Seems like a poor description. The docs even say "Unnecessary qualification can be removed to simplify code and improve readability.".. but this isn't an unnecessary qualification, nor is it being removed.

**Comment by datvm** [comment]
Yeah I don't think it's in code style settings or `.editorconfig` since there seem to be no config option. Probably you have external style/convention guard?

I agree the name is kinda misleading but the suggestion is helpful enough to help me understand the issue so I gave them a pass usually 😅

**Comment by stephentoub** [comment]
Nope, no warnings, no errors, compiles fine.  I expect I'm hitting this: https://github.com/dotnet/roslyn/pull/59075.

**Comment by Youssef1313** [comment]
@stephentoub Are you building from command-line or VS?

I think IDE0002 is broken for command-line builds because it's added in `Features` layer (IDE-only) as opposed to `Analyzers` layer (shared).

**Comment by Youssef1313** [comment]
There is also this 😕 

https://github.com/dotnet/roslyn/blob/9f87b444da9c48a4d492b19f8337339056bf2b95/src/Analyzers/Core/Analyzers/EnforceOnBuildValues.cs#L98

**Comment by jmarolf** [comment]
# Design Meeting Notes

### Discussion

This functionality already exists as part of the simplify-type-name feature but it lacks the ability to be enforces on build or configured to be and error/warning etc. Simplify-type-name does a lot of very expensive things, so we do not want to move all its functionality out to an analyzer. However, we are amenable to carving out an analyzer for this specific bit of functionality.

How should this handle generic parameters? If the derived type had fewer generic parameters, then we should not suggest calling the base type.

For a real-life use case consider this set of test helpers that roslyn uses:
```csharp
namespace Microsoft.CodeAnalysis.Testing
{
	public class CodeFixVerifier<TAnalyzer, TCodeFix, TTest, TVerifier>
           where TAnalyzer : DiagnosticAnalyzer, new()
           where TCodeFix : CodeFixProvider, new()
           where TTest : CodeFixTest<TVerifier>, new()
           where TVerifier : IVerifier, new()
    {
	}
}
```
```csharp
namespace Microsoft.CodeAnalysis.CSharp.Testing
{
    public class CSharpCodeFixVerifier<TAnalyzer, TCodeFix, TVerifier> : CodeFixVerifier<TAnalyzer, TCodeFix, CSharpCodeFixTest<TAnalyzer, TCodeFix, TVerifier>, TVerifier>
        where TAnalyzer : DiagnosticAnalyzer, new()
        where TCodeFix : CodeFixProvider, new()
        where TVerifier : IVerifier, new()
    {
    }
}
```
```csharp
namespace Microsoft.CodeAnalysis.CSharp.Testing.XUnit
{
    public class CodeFixVerifier<TAnalyzer, TCodeFix> : CSharpCodeFixVerifier<TAnalyzer, TCodeFix, XUnitVerifier>
        where TAnalyzer : DiagnosticAnalyzer, new()
        where TCodeFix : CodeFixProvider, new()
    {
    }
}
```
 When consumed 
```csharp
using xunit;
using System.Threading.Tasks;
using Microsoft.CodeAnalysis.CSharp.Testing.XUnit;

namespace MyProjectNamespace
{
    public class MyCodeFixProviderUnitTests
    {
        //No diagnostics expected to show up
        [Fact]
        public async Task TestMethod1()
        {
            var test = @"";

            await CodeFixVerifier<MyAnalyzer, MyCodeFixProvider>.VerifyAnalyzerAsync(test);
        }
	}
}
```
We do not want to suggest simplifying

```csharp
await CodeFixVerifier<MyAnalyzer, MyCodeFixProvider>.VerifyAnalyzerAsync(test);
```
to 
```csharp
await Microsoft.CodeAnalysis.Testing.CodeFixVerifier<MyAnalyzer, MyCodeFixProvider, CSharpCodeFixTest<MyAnalyzer, MyCodeFixProvider, XUnitVerifier>, XUnitVerifier>.VerifyAnalyzerAsync(test);
```

There are existing heuristics in simplify type name that already account for this case we should ensure that they stay in place.

### Conclusion

- Port the functionality from simplify-type-name to its own analyzer with all the existing heuristics in place
- Ensure this rule has its own diagnostic and add a new editorconfig option for this code style rule
- Run this new analyzer/fixer on the dotnet/runtime and dotnet/roslyn repos and see how many false positives are returned
  - If there are no false positives consider having the default severity be "info"
  - Otherwise set the default severity to "silent"
